### PR TITLE
Use SHA1 instead of MD5 for caching

### LINKED ISSUE #2358: FIPS 140-2 compatibility
#### Expected behavior vs actual behavior
We'd like AMS to be compatible with FIPS 140-2 mode. This is currently not working because of the use of MD5 [here](https://github.com/rails-api/active_model_serializers/blob/ab47e983f587e93ece7ab2552537d1475a5eb234/lib/active_model/serializer/concerns/caching.rb#L59).

#### Steps to reproduce
*(e.g., detailed walkthrough, runnable script, example application)*
- install FIPS-compatible OpenSSL binary (this probably differs by OS)
- turn on FIPS mode for OpenSSL with `OpenSSL.fips_mode = true`
- try to invoke an AMS serializer

#### Environment

ActiveModelSerializers Version *(commit ref if not on tag)*: 
0.10.0.rc3, 8981683b9fa8d0bcff0e76b3702d699689b787b4

Output of `ruby -e "puts RUBY_DESCRIPTION"`: 
```
ruby 2.5.6p201 (2019-08-28 revision 67796) [x86_64-linux]
```
OS Type & Version: 
Amazonlinux 2

Integrated application and version *(e.g., Rails, Grape, etc)*: 
Rails 5.2.3

#### Backtrace
*(e.g., provide any applicable backtraces from your application)*
```
md5_dgst.c(82): OpenSSL internal error, assertion failed: Digest MD5 forbidden in FIPS mode!
```
(there is no ruby stack trace, but I verified that `caching.rb` is the culprit)

#### Additonal helpful information
*(e.g., Gemfile.lock, configurations, PR containing a failing test, git bisect results)*

Just replacing MD5 with SHA1 in that file seems to work.
Also note that Rails added FIPS compatibility in 5.2: https://github.com/rails/rails/pull/31651

### PR DESCRIPTION
**Pull request by alexzherdev**
#### Purpose
Using SHA1 as the hashing algorithm allows AMS to be [FIPS 140-2](https://en.wikipedia.org/wiki/FIPS_140-2) compatible.

#### Changes
Only had to change one place where serializer file contents are being hashed.

#### Related GitHub issues
Fixes #2358.

### CODE REVIEW

**Comment by wasifhossain** [comment]
https://github.com/rails-api/active_model_serializers/blob/5d3f56f0a4e3083f34fddcf943e54aed5f5aa5c6/lib/active_model/serializer/concerns/caching.rb#L275

as the digest is added as part of the generated cache key, changing the digest class from MD5 to SHA1 would invalidate all caches.

Hence I would prefer to introduce a **config** that users would use to set their choice of digest function (MD5/SHA). By default, the config would use `MD5` so that users, who do not need to change it, will have their cache intact. Doc should also clearly mention about the cache invalidation fact on changing the digest function.

**Comment by alexzherdev** [comment]
Sure. How about something like a boolean `ActiveModelSerializers.config.use_sha1_digests` just parroting Rails?

**Comment by alexzherdev** [comment]
Updated with the proposed configuration option.

**Comment by bf4** [comment]
pasting rails pr code

railtie

```ruby
      config.after_initialize do
        if app.config.active_support.use_sha1_digests
          ActiveSupport::Digest.hash_digest_class = ::Digest::SHA1
        end
```

**Comment by alexzherdev** [comment]
So with something like
```rb
ActiveModelSerializers.config.use_active_support_digest = true
```
we would read `ActiveSupport::Digest.hash_digest_class`, otherwise `Digest::MD5`?

**Comment by wasifhossain** [comment]
i think `hash_digest_class` config can be set as `Digest::MD5` by default to keep cache intact and only set it to `Digest::SHA1` if the user sets the corresponding boolean field, as pointed in https://github.com/rails-api/active_model_serializers/pull/2361#issuecomment-542884090

**Comment by alexzherdev** [comment]
I'm confused. The code in https://github.com/rails-api/active_model_serializers/pull/2361#issuecomment-542884090 is Rails and `hash_digest_class` is Rails configuration. What should we do in AMS? Should we have a boolean flag, or an option to set the digest class? If someone could post a sample of what you're envisioning, that would be very helpful.

**Comment by wasifhossain** [comment]
My thought was to enable AMS to use a configured digest class across the whole codebase in uniform way. Setting it in railtie would be a good solution in that case.

As AMS is only using it in this specific cache class, I think we can keep it simple for now. This PR looks good in that sense :+1:

I am looking forward to testing the changes soon. In the mean time, can you please add a relevant entry in the CHANGELOG. Thank you!

**Comment by alexzherdev** [comment]
Awesome. Are there any plans to do a release in the near future?

**Comment by wasifhossain** [comment]
we definitely have plans to do that. but we also have to complete some todos before.

in the mean time, you may wish to use the new changes by adding the gem in this way:
```rb
gem 'active_model_serializers', github: 'rails-api/active_model_serializers', branch: '0-10-stable'
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ Breaking changes:
 
 Features:
 
+- [#2361](https://github.com/rails-api/active_model_serializers/pull/2361) Add `ActiveModelSerializers.config.use_sha1_digests` to allow customization of the hashing algorithm used for serializer caching (@alexzherdev)
```
Comment (wasifhossain) [inline CHANGELOG.md]: :+1:

**Comment by wasifhossain** [comment]
would you mind explaining the new config in the DOC too: https://github.com/rails-api/active_model_serializers/blob/0-10-stable/docs/general/configuration_options.md

thanks!

File: docs/general/configuration_options.md [hunk 2]
```diff
@@ -111,6 +111,18 @@ ActiveModelSerializers.config.serializer_lookup_chain.unshift(
...
+
+Determines which hashing algorithm to use internally when caching serializers.
+
+Possible values:
+
+- `true`
+- `false` (default)
+
+When `true`, ActiveModelSerializers will use SHA1. Otherwise, it will default to using MD5.
+Be warned that changing this value may result in serializer caches being invalidated.
```
Comment (wasifhossain) [inline docs/general/configuration_options.md]: :+1:

**Comment by alexzherdev** [comment]
@wasifhossain let me know if there is anything else to handle here 🙇

File: lib/active_model/serializer/concerns/caching.rb [hunk 3]
```diff
@@ -56,7 +56,8 @@ def _cache_digest
         def digest_caller_file(caller_line)
           serializer_file_path = caller_line[CALLER_FILE]
           serializer_file_contents = IO.read(serializer_file_path)
-          Digest::MD5.hexdigest(serializer_file_contents)
+          algorithm = ActiveModelSerializers.config.use_sha1_digests ? Digest::SHA1 : Digest::MD5
```
Comment (bf4) [inline lib/active_model/serializer/concerns/caching.rb]: this is fine

**Pull request merged**